### DISCUSSION TITLE
NextAuth Typescript integration with nextjs13

### DISCUSSION THREAD

**Discussion by Tarbez** [vercel/next.js#42134] [Help]
This is the first time I use, NextAuth and with the major changes of Nextjs 13. I have no idea how to set up nextauth on my project.
I have read the documentation [here](https://next-auth.js.org/getting-started/typescript) 

Somehow not sure how to set it up for nextjs 13. How can I make it work? 
[...nextauth].ts file setup
```
import NextAuth from "next-auth"
import FacebookProvider from "next-auth/providers/facebook";

export const authOptions = {
  // Configure one or more authentication providers
  providers: [
   FacebookProvider({
    clientId: process.env.FACEBOOK_CLIENT_ID,
    clientSecret: process.env.FACEBOOK_CLIENT_SECRET
  }),
  
  ],
}

export default NextAuth(authOptions)
```
Errors so far: 
**Type 'string | undefined' is not assignable to type 'string'.
  Type 'undefined' is not assignable to type 'string'.ts(2322)
oauth.d.ts(83, 5): The expected type comes from property 'clientId' which is declared here on type 'OAuthUserConfig<FacebookProfile>'**

layout.tsx page
```
import { Outfit } from "@next/font/google";

import "../styles/globals.css";
import Header from "./components/Header/Header";
import { SessionProvider } from "next-auth/react";

// Outfit Font
const outfit = Outfit();
export default function RootLayout({
  children,
  pageProps: { session, ...pageProps },
}: {
  children: React.ReactNode;
}) {
  return (
    <SessionProvider session={session}>
      <html>
        <head></head>

        <body lang="en" className={outfit.className}>
          <Header />
          {children}
        </body>
      </html>
    </SessionProvider>
  );
}

```
In the layout file it tells me that Property 'pageProps' does not exist on type '{ children: ReactNode; } 

I would appreciate any help to setup this up.

**Comment by prem5102** [comment]
In next.js 13 App directory does pass page props in Root Layout. use Session Provider without passing session to it. however features also break in next.js 13 when using next auth. please wait until next auth package provide support next.js 13 app directory. in case if you want to next auth to work in next.js 13 refer this [repository](https://github.com/shadcn/taxonomy)

**Comment by Stephenseun09** [comment]
You should probably wrap the only the {children} with SessionProvider... not the entire html

**Comment by ahamedzoha** [comment]
add the line `"use client"` to the top

and use the SessionProvider **without** the session prop as such:

```
"use client"
...
<html>
    <head></head>
    <SessionProvider>
      <body lang="en" className={outfit.className}>
            <Header />
            {children}
       </body>
      </SessionProvider>
</html>

```


You could also create a provider wrapper and wrap the children prop. Then you could use it here without using the `"use client"` directive in the `layout.tsx`

**Comment by olesuv** [comment]
I had the same problem. On setup section of using NextJS I selected `~/src/app` architecture type. After I made this setup for NextAuth:
1. Made this file `~/src/app/api/[...nextauth]/route.ts`:
```ts
import { authConfig } from "@/app/lib/auth";
import NextAuth from "next-auth/next";

const handler = NextAuth(authConfig);

export { handler as GET, handler as POST };
```
2. Added file in which should be config for NextAuth `~/src/app/lib/auth.ts`:
```ts
import { NextAuthOptions } from "next-auth";

import GoogleProvider from "next-auth/providers/google";
import GitHubProvider from "next-auth/providers/github";

export const authConfig: NextAuthOptions = {
  session: {
    strategy: "jwt",
  },

  providers: [
    GoogleProvider({
      clientId: process.env.GOOGLE_CLIENT_ID as string,
      clientSecret: process.env.GOOGLE_CLIENT_SECRET as string,
    }),

    GitHubProvider({
      clientId: process.env.GITHUB_ID as string,
      clientSecret: process.env.GITHUB_SECRET as string,
    }),
  ],
};
```
3. As @ahamedzoha mentioned in file `~/src/app/layout.tsx` should be added:
- `<SessionProvider>` before and after `<body>...</body>`.
- `"use client"` at the top of the file, because that's React client component. 

## Example how to use `useSession()` in project:
- Page:
```ts
import TestAuth from "@/components/home/testAuth";

export default async function Home() {
  return (
    <div className="flex flex-row w-full text-justify text-xl font-semibold">
      <TestAuth />
    </div>
  );
}
```
- Component:
```ts
"use client";

import { redirect } from "next/navigation";
import { useSession } from "next-auth/react";

export default function TestAuth(props: any) {
  const { data: session, status } = useSession();

  if (status === "authenticated") {
    return <span>Signed in as {session.user?.name}</span>;
  } else if (status === "loading") {
    return <span className="text-teal-800">Loading...</span>;
  } else {
    redirect("/");
  }
}
```

## Credentials/used sources: 
- @ahamedzoha 
- @ipenywis and his [NextAuth guide](https://youtu.be/AbUVY16P4Ys).
